File: CPP/ByRound/1971/1971_E_Find_the_Car.cpp
// Problem: CF 1971 E - Find the Car
// https://codeforces.com/contest/1971/problem/E

/*
E. Find the Car

Purpose:
This program processes multiple test cases where each case defines a car's travel on a number line from 0 to n.
The car's position and time are known at k+1 checkpoints (including start and end).
Between any two points, the speed is constant. For each query, we determine the time it takes to reach a certain distance d.

Algorithms:
- Binary search using std::upper_bound for efficient lookup of the correct segment.
- Linear interpolation between known time points for computing intermediate times.

Time Complexity: O(k log k + q log k) per test case
Space Complexity: O(k) per test case

*/
#include <algorithm>
#include <cstdio>
#include <vector>
typedef long long ll;

int main() {
  ll t;
  scanf("%lld", &t);
  while (t--) {
    ll n, k, q;
    scanf("%lld %lld %lld", &n, &k, &q);
    
    // Store positions and times
    std::vector<ll> a(k + 1, 0), b(k + 1, 0);
    
    // Read positions
    for (ll p = 1; p <= k; p++) {
      scanf("%lld", &a[p]);
    }
    
    // Read times
    for (ll p = 1; p <= k; p++) {
      scanf("%lld", &b[p]);
    }
    
    // Process queries
    for (ll p = 0; p < q; p++) {
      ll d;
      scanf("%lld", &d);
      
      // If query point is exactly at the final destination
      if (d == n) {
        printf("%lld ", b.back());
        continue;
      }
      
      // Find the index of the first element in 'a' greater than 'd'
      ll idx = std::upper_bound(a.begin(), a.end(), d) - a.begin();
      
      // Interpolate time using the formula:
      // t = t1 + (t2 - t1) * (d - d1) / (d2 - d1)
      printf("%lld ", b[idx - 1] + (b[idx] - b[idx - 1]) * (d - a[idx - 1]) /
                                       (a[idx] - a[idx - 1]));
    }
    puts("");
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/1971/1971_E_Find_the_Car.py
# Problem: CF 1971 E - Find the Car
# https://codeforces.com/contest/1971/problem/E

"""
Algorithm: Binary Search + Linear Interpolation

Purpose: 
This program processes multiple test cases to determine the time taken for a car 
to reach a specific point on a number line, given known positions and times at 
intermediate checkpoints. The car moves at a constant speed between consecutive points.

Approach:
1. For each query (point d), find the interval [a[l], a[l+1]] that contains d using binary search.
2. Compute the time to reach d using linear interpolation:
   - Time to reach a[l] is b[l]
   - Speed between a[l] and a[l+1] is (b[l+1] - b[l]) / (a[l+1] - a[l])
   - Time to travel from a[l] to d is (d - a[l]) * speed
   - Total time = b[l] + (d - a[l]) * (b[l+1] - b[l]) / (a[l+1] - a[l])

Time Complexity: O(q * log k) per test case
Space Complexity: O(k) per test case for storing a and b arrays

"""
import sys

input = sys.stdin.readline

ints = lambda: list(map(int, input().split()))

t = int(input())
for _ in range(t):
    n, k, q = ints()
    a = ints()
    b = ints()
    a = [0] + a  # Prepend 0 to a and b for easier index handling
    b = [0] + b

    for _ in range(q):
        d = int(input())

        # Binary search to find the largest index l such that a[l] <= d
        l = 0
        r = len(a) - 1

        while l <= r:
            m = (l + r) // 2
            if a[m] > d:
                r = m - 1
            else:
                l = m + 1
        l -= 1  # l is now the largest index with a[l] <= d

        # Calculate time to reach point d
        ans = b[l]  # Time to reach a[l]
        if l < len(a) - 1:  # If not at the final point
            # Add time to travel from a[l] to d using linear interpolation
            ans += ((d - a[l]) * (b[l + 1] - b[l])) // (a[l + 1] - a[l])
        ans = int(ans)
        print(ans, end=" ", flush=False)
    print()


# https://github.com/VaHiX/CodeForces/